src/Simplex/Messaging/Transport.hs

```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
           T.serverSupported = supportedParameters
         }
 
+mkTLSClientParams :: HostName -> ServiceName -> C.CertificateHash -> T.ClientParams
+mkTLSClientParams host port certificateHash = do
+  let p = B.pack port
+  (T.defaultParamsClient host p)
+    { T.clientShared = def,
+      T.clientHooks = def {T.onServerCertificate = \_ _ _ -> validateCertificateChain certificateHash host p},
+      T.clientSupported = supportedParameters
+    }
+
+validateCertificateChain :: C.CertificateHash -> HostName -> ByteString -> X.CertificateChain -> IO [XV.FailedReason]
+validateCertificateChain _ _ _ (X.CertificateChain []) = pure [XV.EmptyChain]
+validateCertificateChain expectedHash host port cc@(X.CertificateChain sc@[cert]) = do
+  let certificateHash = C.certificateHash . X.getCertificate $ cert
+      fr = checkHash certificateHash
+  -- putStrLn $ "certificate hash: " <> (B.unpack . encode . C.unCertificateHash) certificateHash
+  -- putStrLn $ "expected hash: " <> (B.unpack . encode . C.unCertificateHash) expectedHash
+  if null fr
+    then x509validate
+    else pure fr
+  where
+    checkHash :: C.CertificateHash -> [XV.FailedReason]
+    checkHash hash
+      | hash == expectedHash = []
+      | otherwise = [XV.UnknownCA]
+    x509validate :: IO [XV.FailedReason]
+    x509validate = XV.validate X.HashSHA256 hooks checks certStore cache serviceID cc
+      where
+        hooks :: XV.ValidationHooks
+        hooks = XV.defaultHooks
+        checks :: XV.ValidationChecks
+        checks = XV.defaultChecks {XV.checkLeafV3 = False} -- TODO create v3 certificates? https://stackoverflow.com/a/18242720
+        certStore :: XS.CertificateStore
+        certStore = XS.makeCertificateStore sc
+        cache :: XV.ValidationCache
+        cache = XV.exceptionValidationCache [] -- we don't store certificate Fingerprints on clients
+        serviceID :: XV.ServiceID
+        serviceID = (host, port)
+validateCertificateChain _ _ _ (X.CertificateChain (_ : _ : _)) = pure [XV.AuthorityTooDeep]
+
+supportedParameters :: T.Supported
+supportedParameters =
+  def
+    { T.supportedVersions = [T.TLS12],
+      T.supportedCiphers = [TE.cipher_ECDHE_ECDSA_CHACHA20POLY1305_SHA256],
+      T.supportedHashSignatures = [(T.HashIntrinsic, T.SignatureEd448), (T.HashIntrinsic, T.SignatureEd25519)],
+      T.supportedSecureRenegotiation = False,
+      T.supportedGroups = [T.X448, T.X25519]
+    }
+
+getCertificateHash :: FilePath -> IO ByteString
+getCertificateHash certificateFile = do
+  x509 <- SX.readSignedObject certificateFile
+  let certificate = X.getCertificate (head x509) -- we should have only one certificate
+  pure $ (encode . C.unCertificateHash) (C.certificateHash certificate)
+
```

Comment:
please don't move things around to create unnecessary diff - the location of code in the file is not important, no need to adhere to the sections anywhere other than in module exports - functions can be anywhere.

Please return them where they were.